### PR TITLE
Typesafe commands

### PR DESCRIPTION
**Pull request by jeffsee55**
- Streamlines the dev server workflow in a few ways
  - Removes the express server in favor of middleware on the Vite server
  - Removes the altair endpoint in favor of a GraphiQL playground as part of the Admin. This was a necessary change due to the express server going away
  - Removes the prebuild step for user-provided config
  - Removes the awkward setup we had in `@tinacms/app`. Instead, this package is now plain Vite app which exports an `index.html`, there is no build as that happens as part of the `dev` and `build` processes.
- `@tinacms/graphql` no longer writes SDK files to the filesystem, it just returns their strings to the cli
- When content is changed from the files directly, they're updated in the database immediately

### CODE REVIEW

File: packages/@tinacms/app/index.html
Comment (logan-anderson) [inline packages/@tinacms/app/index.html:12]: Was this here before.

```
window.require = () => {}
````
? 

Seems like it _might_ cause issues 🤔  ?

File: packages/@tinacms/cli/src/logger/index.ts
Comment (logan-anderson) [inline packages/@tinacms/cli/src/logger/index.ts:67]: Should this be the configured port instead of 4001?

File: packages/@tinacms/cli/src/logger/index.ts
Comment (logan-anderson) [inline packages/@tinacms/cli/src/logger/index.ts:98]: Should we be using our Logger module here?

File: packages/@tinacms/cli/src/next/commands/build-command/server/index.ts
Comment (logan-anderson) [inline packages/@tinacms/cli/src/next/commands/build-command/server/index.ts:15]: Looks like this functions is declared in 2-3 places. Maybe it should be unified into one place?

File: packages/@tinacms/cli/src/next/commands/dev-command/html.ts
Comment (logan-anderson) [inline packages/@tinacms/cli/src/next/commands/dev-command/html.ts:11]: Should this be the port passed in `--port` and not `4001`

File: packages/@tinacms/cli/src/next/config-manager.ts
Comment (logan-anderson) [inline packages/@tinacms/cli/src/next/config-manager.ts:46]: I think we have a normalize path function that _could_ be used here

File: packages/@tinacms/cli/src/next/config-manager.ts
Comment (logan-anderson) [inline packages/@tinacms/cli/src/next/config-manager.ts:57]: Is this needed? Could we just used the default behaviour of dotenv?

File: packages/@tinacms/scripts/src/index.ts
Comment (logan-anderson) [inline packages/@tinacms/scripts/src/index.ts:523]: Remove ?

File: packages/@tinacms/cli/src/next/commands/dev-command/server/index.ts
Comment (logan-anderson) [inline packages/@tinacms/cli/src/next/commands/dev-command/server/index.ts:149]: I think this link should be `/index.html#/graphql`

**Comment by logan-anderson** [comment]
Couple of questions regrading the graphQL playground.

- Is there any reason why  can't or don't serve it from localhost:4001/ anymore?
- The link when your run `tinacms dev` still points you to loclahost:4001 (Maybe it should not link you to your dev URL)
- Since we are rendering the GraphQL playground now could we wrap it in our own navigation and link to it in the sidebar

File: packages/@tinacms/cli/src/next/commands/build-command/index.ts
Comment (logan-anderson) [inline packages/@tinacms/cli/src/next/commands/build-command/index.ts:111]: Maybe we could add the admin to this list as well? (Seems to be a common place of confusion)

**Comment by jeffsee55** [comment]
> Is there any reason why can't or don't serve it from localhost:4001/ anymore?

Altair had a built-in express handler, and we're not using express with this PR, we could probably wire it up still with SSR of some sort, but at the same time I'd probably rather this playground live closer to the CMS SPA anyway. We can do more with it in the future.

> The link when your run tinacms dev still points you to loclahost:4001 (Maybe it should not link you to your dev URL)

Yeah, will fix. This is a still a draft FYI

> Since we are rendering the GraphQL playground now could we wrap it in our own navigation and link to it in the sideba

Yeah agreed, there's some CSS conflicts right now, even with the way I have it working. But ultimately I'd like it to feel more part of the CMS, and we can provide some buttons to populate the queries with our autogenerated queries, which is something we couldn't really do before.

File: packages/@tinacms/cli/src/next/config-manager.ts
Comment (jeffsee55) [reply packages/@tinacms/cli/src/next/config-manager.ts:57]: I'm happy for it be a bit more explicit, if we need to open it up we can for some reason. This has the advantage of working when you're running the CLI from the monorepo and using `--rootPath`. So if it causes issues we can just use this conditionally.

File: packages/@tinacms/cli/src/logger/index.ts
Comment (jeffsee55) [reply packages/@tinacms/cli/src/logger/index.ts:52]: I'm hoping to combine these in the near future. It's all in the logger module for now regardless.

File: packages/@tinacms/cli/src/index.ts
Comment (jeffsee55) [inline packages/@tinacms/cli/src/index.ts:28]: @kldavis4 I think we're ready to support the `tina` folder since this PR https://github.com/tinacms/tinacms/pull/3563/files. But I don't recall exactly where we left off. This command is a codemod to move the folder, but if you have a `tina` folder things should work regardless.

**Comment by logan-anderson** [comment]
Also getting this error when running the basic starter.

```
✘ [ERROR] [plugin vite:dep-scan] ENOENT: no such file or directory, open '/Users/logananderson-forestry-mac/dev/tinacms/tinacms/examples/basic-iframe/.tina/__generated__/types.ts'

    ../../node_modules/.pnpm/esbuild@0.16.17/node_modules/esbuild/lib/main.js:1426:27:
      1426 │         let result = await callback({
           ╵                            ^

    at Object.openSync (node:fs:594:3)
    at Object.readFileSync (node:fs:462:35)
    at file:///Users/logananderson-forestry-mac/dev/tinacms/tinacms/node_modules/.pnpm/vite@4.1.4/node_modules/vite/dist/node/chunks/dep-ca21228b.js:42096:37
    at requestCallbacks.on-load (/Users/logananderson-forestry-mac/dev/tinacms/tinacms/node_modules/.pnpm/esbuild@0.16.17/node_modules/esbuild/lib/main.js:1426:28)
    at handleRequest (/Users/logananderson-forestry-mac/dev/tinacms/tinacms/node_modules/.pnpm/esbuild@0.16.17/node_modules/esbuild/lib/main.js:727:19)
    at handleIncomingPacket (/Users/logananderson-forestry-mac/dev/tinacms/tinacms/node_modules/.pnpm/esbuild@0.16.17/node_modules/esbuild/lib/main.js:749:7)
    at Socket.readFromStdout (/Users/logananderson-forestry-mac/dev/tinacms/tinacms/node_modules/.pnpm/esbuild@0.16.17/node_modules/esbuild/lib/main.js:677:7)
    at Socket.emit (node:events:513:28)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)

  This error came from the "onLoad" callback registered here:

    ../../node_modules/.pnpm/esbuild@0.16.17/node_modules/esbuild/lib/main.js:1287:20:
      1287 │       let promise = setup({
           ╵                     ^

    at setup (file:///Users/logananderson-forestry-mac/dev/tinacms/tinacms/node_modules/.pnpm/vite@4.1.4/node_modules/vite/dist/node/chunks/dep-ca21228b.js:42092:19)
    at handlePlugins (/Users/logananderson-forestry-mac/dev/tinacms/tinacms/node_modules/.pnpm/esbuild@0.16.17/node_modules/esbuild/lib/main.js:1287:21)
    at buildOrServeImpl (/Users/logananderson-forestry-mac/dev/tinacms/tinacms/node_modules/.pnpm/esbuild@0.16.17/node_modules/esbuild/lib/main.js:974:5)
    at Object.buildOrServe (/Users/logananderson-forestry-mac/dev/tinacms/tinacms/node_modules/.pnpm/esbuild@0.16.17/node_modules/esbuild/lib/main.js:780:5)
    at /Users/logananderson-forestry-mac/dev/tinacms/tinacms/node_modules/.pnpm/esbuild@0.16.17/node_modules/esbuild/lib/main.js:2132:17
    at new Promise (<anonymous>)
    at Object.build (/Users/logananderson-forestry-mac/dev/tinacms/tinacms/node_modules/.pnpm/esbuild@0.16.17/node_modules/esbuild/lib/main.js:2131:14)
    at build (/Users/logananderson-forestry-mac/dev/tinacms/tinacms/node_modules/.pnpm/esbuild@0.16.17/node_modules/esbuild/lib/main.js:1978:51)
    at scanImports (file:///Users/logananderson-forestry-mac/dev/tinacms/tinacms/node_modules/.pnpm/vite@4.1.4/node_modules/vite/dist/node/chunks/dep-ca21228b.js:41785:11)

  The plugin "vite:dep-scan" was triggered by this import

    ../../packages/@tinacms/app/src/Playground.tsx:7:24:
      7 │ import { queries } from 'CLIENT_IMPORT'
```

File: packages/@tinacms/cli/src/next/vite/index.ts
Comment (logan-anderson) [inline packages/@tinacms/cli/src/next/vite/index.ts:8]: Might be nitpicky, but can the `createConfig` function be take a single object argument instead of a list of 6 args.


I find
```ts
 const config = await createConfig(
    configManager,
    database,
    apiURL,
    [],
    noSDK,
    true
  )
```

Hard to understand.

File: packages/@tinacms/cli/src/next/commands/dev-command/index.ts
Comment (logan-anderson) [inline packages/@tinacms/cli/src/next/commands/dev-command/index.ts:55]: Two things here

1. I think this should log even if verbose is not set. (Since the default is false)

Leaving someone with just "Error occured during tinacms dev" by defualt make it really heard to debug IMO. 

2. Are we stopping the sub process when TinaCMS dev has an error? (I just got into a state where the sub process was running but TinaCMS dev errored)

**Comment by logan-anderson** [comment]
Looks good to me. Just two more things I noticed

1. The GraphQL playground is awesome!
- Can we link to it in the playground from the sidebar? (Does not have to block this PR)
- Is there anyway to make the popups work? (It is ok if not but I do find it awkward right now without then)
    - Maybe we could have another entry point to out vite app that build its it totally separate to something like /admin/graphql.html


2. I noticed that this removes the "isomorphicGitBridge" from the  `_schema.json` I know this does not matter locally does it matter in TinaCloud @kldavis4 ?

Overall looking really good

**Comment by kldavis4** [comment]
> 2. I noticed that this removes the "isomorphicGitBridge" from the  `_schema.json` I know this does not matter locally does it matter in TinaCloud @kldavis4 ?

Tina cloud is no longer expecting any features flags (including the isomorphicGitBridge)

**Comment by jamespohalloran** [comment]
Testing this with a PR deploy, I'm getting this new error on tina init
Screen Shot 2023-03-14 at 1 59 15 PM

**Comment by jeffsee55** [comment]
> Testing this with a PR deploy, I'm getting this new error on tina init

I'm guessing this was a missed merge conflict. I'll take a look, thanks.

File: packages/@tinacms/toolkit/src/components/media/breadcrumb.tsx
Comment (jeffsee55) [reply packages/@tinacms/toolkit/src/components/media/breadcrumb.tsx:13]: Fixed, thanks.

**Comment by jeffsee55** [comment]
@jamespohalloran that's been fixed, thanks.

> So does this need to be uncommented after this is merged? or before?

This can be uncommented once we know things work on Tina Cloud

**Comment by jeffsee55** [comment]
> Can we link to it in the playground from the sidebar? (Does not have to block this PR)

Yeah I want to let the dust settle first. Mainly because it's a bit buggy with the CSS issues

> Is there anyway to make the popups work? (It is ok if not but I do find it awkward right now without then)

I did this initially, but we have our auth wrapper around things, and this actually works in the hosted environment too, which is really handy and I'd like to keep it around. So I think the proper fix is to clean up our admin fixed positioning stuff.

**Comment by logan-anderson** [comment]
@jeffsee55 I ran `yarn add tinacms@typesafe-clipanion @tinacms/cli@typesafe-clipanion`

to the basic starter (created with create-tina-app) and got this error in the terminal
```
(!) Could not auto-determine entry point from rollupOptions or html files and there are no explicit optimizeDeps.include patterns. Skipping dependency pre-bundling.
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
event - compiled client and server successfully in 1288 ms (173 modules)
wait  - compiling / (client and server)...
event - compiled client and server successfully in 463 ms (237 modules)
Transform failed with 1 error:
<stdin>:8:0: ERROR: Unexpected "*"
2:17:46 p.m. [vite] Internal server error: Transform failed with 1 error:
<stdin>:8:0: ERROR: Unexpected "*"
  Plugin: transform-tsx
  File: /Users/logananderson-forestry-mac/dev/playground/tina-test-march-15/node_modules/@tinacms/app/src/index.css
      at failureErrorWithLog (/Users/logananderson-forestry-mac/dev/playground/tina-test-march-15/node_modules/esbuild/lib/main.js:1566:15)
      at /Users/logananderson-forestry-mac/dev/playground/tina-test-march-15/node_modules/esbuild/lib/main.js:805:29
      at responseCallbacks.<computed> (/Users/logananderson-forestry-mac/dev/playground/tina-test-march-15/node_modules/esbuild/lib/main.js:671:9)
      at handleIncomingPacket (/Users/logananderson-forestry-mac/dev/playground/tina-test-march-15/node_modules/esbuild/lib/main.js:726:9)
      at Socket.readFromStdout (/Users/logananderson-forestry-mac/dev/playground/tina-test-march-15/node_modules/esbuild/lib/main.js:647:7)
      at Socket.emit (node:events:513:28)
      at addChunk (node:internal/streams/readable:315:12)
      at readableAddChunk (node:internal/streams/readable:289:9)
      at Socket.Readable.push (node:internal/streams/readable:228:10)
      at Pipe.onStreamRead (node:internal/stream_base_commons:190:23)
Transform failed with 2 errors:
<stdin>:1:5: ERROR: Expected "class" after TypeScript decorator but found "-"
<stdin>:1:11: ERROR: Expected ";" but found "{"
``` 

and this error in the browser when I went to the /admin page.


Screenshot 2023-03-15 at 2 18 29 PM

**Comment by jeffsee55** [comment]
Thanks @logan-anderson. This is fixed in the latest commit, and the PR has been redeployed

**Comment by logan-anderson** [comment]
This can probably move to a separate issue but right now if there is an error during a re-build the entire dev process crashes. Previously it still run with the old version.

See https://www.loom.com/share/fd304f54fbab4a56a6013d22ed8e4789

**Review by logan-anderson — APPROVED** [review]
I will leave it up to you weather or not you want to move https://github.com/tinacms/tinacms/pull/3659#issuecomment-1470997253 to a seperate issue or fix here but other then that it looks good.

**Comment by steffkes** [comment]
Hey there, I'm late to the party - sorry for that! Never had my hands on tinacms before the migration from forest.io

While attempting the migration, an issued occurred which - I think - originated in this pull request. Previously there was an Option to expose the Vite server externally (introduced by #3484) - which doesn't seem to be there anymore.

When we run `tinacms dev --verbose -c nuxt` (which is configured to listen to Port `3000`), I have the following processes on those open ports in the container:

```
# netstat -nltp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.11:43609        0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:3000            0.0.0.0:*               LISTEN      68/node
tcp        0      0 127.0.0.1:4001          0.0.0.0:*               LISTEN      31/node
tcp6       0      0 :::9000                 :::*                    LISTEN      31/node
```

to toy around with and potentially get tinacms up and running, we need the `4001` port configured the same way. I see how the default `4001` port gets passed down in the `dev-command` ([reference](https://github.com/tinacms/tinacms/blob/a982b105d685fc343b989462c882d56c72a04007/packages/%40tinacms/cli/src/next/commands/dev-command/index.ts#L195)) - but the `host` for [Vite needs `server.host` to be passed as part of their Configuration Object](https://vitejs.dev/config/server-options.html#server-host).

So far for my analysis - any suggestions on how to take it further @jeffsee55? I'm not very familiar with tinacms, but would like to help .. if you let me know _how_. 

Thanks
-Stefan

**Comment by steffkes** [comment]
> EDIT: does this have anything to do with anything?
> 
> https://github.com/tinacms/tinacms/blob/a982b105d685fc343b989462c882d56c72a04007/packages/%40tinacms/app/src/App.tsx#L7-L9

What's your thinking here @wommy? If there would be any error from this, I'd expect it to occur once you request the page in a browser .. but we're talking about the server listening only on a local port that is not exposed to the outside.

> even tried passing in `tina/config.ts`
> 
> ```ts
> export default defineConfig({
>   build: {
>     host: true,
>   },
> }
> ```

this is what #3484 added as functionality, but I don't see a code path anymore where this gets passed down the the vite config as it did before?

**Comment by wommy** [comment]
nah sorry I was braindead from trying to debug this for a few hours, reaching at straws

but also shouldnt it be 
```ts
export default defineConfig({
  server: {
    host: true,
  },
}
```
https://vitejs.dev/config/server-options.html#server-host

**Comment by logan-anderson** [comment]
@wommy @steffkes You can not currently change the vite config from DefineConfig. 

So at the momant 
```
export default defineConfig({
  server: {
    host: true,
  },
}
```

Does not do anything. 

I can make a PR do add back `host: true`

**Pull request merged**
